File: pipeline/frontend/src/types/science.ts
/**
 * The scientific schema, exactly as the extraction service serialises it
 * (see `extraction/app/schemas/science.py`).
 *
 *   experiments ─┬─ experiment_ingredients ── ingredients
 *                └─ measurements ── indicators   (+ evidence)
 *
 * These are the rows the LLM produced, read verbatim. There is no second, reshaped copy of
 * them any more: what the reviewer sees here is what every downstream consumer reads.
 */

export interface Ingredient {
  id: number
  ingredient_name: string
  functional_class: string
  /** Biological origin of the ingredient, not the paper it came from. */
  source: string
}

export interface Indicator {
  id: number
  /** What the paper calls it, e.g. "Total viable count". This is the label to display. */
  indicator_name: string
  /** The category it belongs to: "microbial" or "chemical". Not a display name. */
  indicator_type: string
  indicator_unit: string
  indicator_threshold: number | null
}

/** One ingredient as it is used in a specific experiment, with its concentration. */
export interface ExperimentIngredient {
  ingredient_id: number
  ingredient_name: string
  functional_class: string
  source: string
  /** Null when the paper names the additive without stating how much was used. */
  concentration: number | null
  concentration_unit: string | null
}

/** One value at one (day, indicator) for one experiment. */
export interface Measurement {
  day: number
  indicator_id: number
  indicator_name: string
  indicator_type: string
  indicator_unit: string
  indicator_threshold: number | null
  indicator_value: number
  /**
   * How many separately-printed readings were averaged into this value; 1 when the paper
   * printed a single value or its own mean. Replaces `value_is_approximate`, which
   * recorded a provenance fact the evidence rows already carry.
   */
  replicates: number
}

export interface BoundingBox {
  x1: number
  y1: number
  x2: number
  y2: number
}

export interface Evidence {
  id: number
  entity_type: string
  entity_key: Record<string, unknown>
  field_name: string
  page_number: number | null
  /** "prose" | "table" | "figure" */
  source_type: string
  source_label: string | null
  exact_text: string | null
  /**
   * How the value was arrived at: "stated", "derived" or "inferred". The last two carry a
   * `rationale`, so a computed value is never read as one the paper stated outright.
   */
  method: string
  rationale: string | null
  bounding_box: BoundingBox | null
  /** Scored by the pipeline against the source text, never supplied by the model. */
  confidence: number | null
  figure_series: string | null
  x_axis_value: number | null
  y_axis_value: number | null
  value_is_approximate: boolean
  is_chart_derived: boolean
}

export interface ExperimentSummary {
  id: number
  paper_id: number
  meat_matrix: string
  /**
   * Null when the paper never describes this arm's handling. Whether it addressed it at
   * all is answered by an evidence span for `field_name === "treatment"`, not by this.
   */
  treatment: string | null
  /** Mass in grams of one experimental sample unit. */
  weight_g: number | null
  ingredients: ExperimentIngredient[]
  measurement_count: number
  created_at: string | null
}

export interface ExperimentDetail {
  id: number
  paper_id: number
  meat_matrix: string
  treatment: string | null
  weight_g: number | null
  created_at: string | null
  ingredients: ExperimentIngredient[]
  measurements: Measurement[]
  evidence: Evidence[]
}

File: pipeline/processing/app/schemas/science.py
"""Request and response models for the scientific schema."""

from __future__ import annotations

from datetime import datetime
from typing import Any, Optional

from pydantic import BaseModel, ConfigDict


class IngredientOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    ingredient_name: str
    functional_class: str
    source: str


class IndicatorOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    #: What the paper calls it, e.g. "Total viable count". This is the label to show.
    indicator_name: str
    #: The category it belongs to: "microbial" or "chemical". Not a display name.
    indicator_type: str
    indicator_unit: str
    indicator_threshold: Optional[float] = None


class IndicatorUpdate(BaseModel):
    """The threshold only.

    Name, type and unit are extraction output and identify the row -- name and unit form
    its uniqueness constraint -- so editing them here would either collide with another
    indicator or silently relabel every measurement pointing at this one.
    """

    indicator_threshold: Optional[float] = None


class ExperimentIngredientOut(BaseModel):
    ingredient_id: int
    ingredient_name: str
    functional_class: str
    source: str
    #: Null when the paper names the additive without stating how much was used.
    concentration: Optional[float] = None
    concentration_unit: Optional[str] = None


class MeasurementOut(BaseModel):
    day: int
    indicator_id: int
    indicator_name: str
    indicator_type: str
    indicator_unit: str
    indicator_threshold: Optional[float] = None
    indicator_value: float
    #: How many separately-printed readings were averaged into this value; 1 when the paper
    #: printed a single value or its own mean.
    replicates: int = 1


class BoundingBox(BaseModel):
    x1: float
    y1: float
    x2: float
    y2: float


class EvidenceOut(BaseModel):
    id: int
    entity_type: str
    entity_key: dict[str, Any]
    field_name: str
    page_number: Optional[int] = None
    source_type: str
    source_label: Optional[str] = None
    exact_text: Optional[str] = None
    #: How the value was arrived at: "stated", "derived" or "inferred". The last two carry
    #: a `rationale`, so a computed value is never read as one the paper stated outright.
    method: str
    rationale: Optional[str] = None
    bounding_box: Optional[BoundingBox] = None
    #: Scored by the pipeline against the source text, never supplied by the model.
    confidence: Optional[float] = None
    figure_series: Optional[str] = None
    x_axis_value: Optional[float] = None
    y_axis_value: Optional[float] = None
    value_is_approximate: bool = False
    is_chart_derived: bool = False


class ExperimentSummaryOut(BaseModel):
    id: int
    paper_id: int
    meat_matrix: str
    #: Null when the paper never describes this arm's handling. Whether it addressed it at
    #: all is answered by an evidence span for `field_name="treatment"`, not by this field.
    treatment: Optional[str] = None
    weight_g: Optional[float] = None
    ingredients: list[ExperimentIngredientOut]
    measurement_count: int
    created_at: Optional[datetime] = None


class ExperimentDetailOut(BaseModel):
    id: int
    paper_id: int
    meat_matrix: str
    treatment: Optional[str] = None
    weight_g: Optional[float] = None
    created_at: Optional[datetime] = None
    ingredients: list[ExperimentIngredientOut]
    measurements: list[MeasurementOut]
    evidence: list[EvidenceOut]
